src/app/initialisation/run.js

```diff
@@ -57,42 +59,45 @@ function initCoreModules(mediator) {
  * Monitoring any changes in the state to verify that the user
  * @param $rootScope
  * @param $state
- * @param $q
- * @param mediator
  * @param userClient
  */
-function verifyLoginOnStateChange($rootScope, $state, $q, mediator, userClient) {
-  var initPromises = [];
-  var initListener = mediator.subscribe('promise:init', function(promise) {
-    initPromises.push(promise);
-  });
-  mediator.publish('init');
-  var all = (initPromises.length > 0) ? $q.all(initPromises) : $q.when(null);
-  all.then(function() {
-    $rootScope.ready = true;
-    mediator.remove('promise:init', initListener.id);
-    return null;
-  });
+function verifyLoginOnStateChange($rootScope, $state, userClient) {
+
+  $rootScope.ready = false;
+
+  function clearAndRedirectToLogin() {
+    userClient.clearSession().then(function() {
+      $state.go('app.login');
+    });
+  }
 
   $rootScope.$on('$stateChangeStart', function(e, toState, toParams) {
 
-    function clearAndRedirectToLogin() {
-      userClient.clearSession().then(function() {
-        $rootScope.toState = toState;
-        $rootScope.toParams = toParams;
-        $state.go('app.login');
-      });
+    //If the session is valid, or accessing the login page, then the transition is valid
+    if (isAuthorised || toState.name === "app.login") {
+      $rootScope.ready = true;
+      return;
     }
 
-    if (toState.name !== "app.login") {
-      userClient.verifySession().then(function(validSession) {
-        //If the session is not valid, clear the user data and redirect to login
-        if (!validSession) {
-          e.preventDefault();
-          clearAndRedirectToLogin();
+    //Preventing the route from executing. Verifying that the session is valid first.
+    e.preventDefault();
+
+    userClient.verifySession().then(function(validSession) {
```

Comment:
[Question] Not related to this PR. Why we need to verify session explicitly?
In typical api when session is outdated client will get 401 with next request and this will require user to login again. Why our case is different? Is that a sync problem? For some critical security applications going to background usually logs out user so when he resumes application user need to login or enter pin again.